pkg/apis/core/types.go

```diff
@@ -4142,11 +4142,48 @@ type NodeAddressType string
 
 // These are valid values of node address type
 const (
-	NodeHostName    NodeAddressType = "Hostname"
-	NodeExternalIP  NodeAddressType = "ExternalIP"
-	NodeInternalIP  NodeAddressType = "InternalIP"
-	NodeExternalDNS NodeAddressType = "ExternalDNS"
+	// NodeHostName identifies the node's canonical hostname
```

Comment:
"In general, this should match the name of the node but is not strictly required." 

^ I don't know if this the last part is actually true since I've never seen cases where the hostname doesn't match.

Reply:
This is a big terminology mess in k8s, but: Node name != `os.Hostname()` (aka `uname -n`) != DNS entry
It gets worse, because (eg) `kubelet --hostname-override` actually sets the node name, not the hostname. And other examples.

On my on-prem cluster, for example: None of the hosts have entries in DNS, and I use /etc/machine-id as the node name (ie: `kubelet --hostname-override=%m` in systemd config).  `uname -n` reports "localhost" (CoreOS/flatcar out-of-the-box default) on all of them.  Kubernetes works fine in this setup, and honestly I think it's superior to trying to align DNS primary keys for no reason.  Other common examples include OpenStack, where there is (was?) commonly no DNS entry for hosts but the node-name == openstack instance name, and AWS with DHCP options, where the hostname (domain suffix) might not match the generated node name.  Basically the only place where node name == uname -n == DNS is GCE, so it's not surprising that k8s had this confused initially :stuck_out_tongue:

Reply:
I think that canonical hostname is ok, we are not going to solve a 30 years old problem ;)

Reply:
> I think that canonical hostname is ok

Where 'canonical hostname' means `os.Hostname()` or "DNS entry" (or `Node.metadata.name`) ?  Please pick one, because I don't think 'canonical hostname' is at all clear in k8s context.  I'm not asking you to solve anything except this comment :)

I suggest `os.Hostname()` (aka `uname -n`) - since we already have NodeInternalDNS/NodeExternalDNS for "DNS Entry", and we obviously already have the `Node` name elsewhere in the `Node` object.

Reply:
is it clear in a typical unix system? :)

Reply:
Current behavior is:
- AWS: primary DNS name
- Azure: `node.Name`
- GCE: primary DNS name
- OpenStack: "hostname". not clear without further digging
- vSphere: `os.Hostname()`
- "bare metal": `os.Hostname()` or command-line override

I had thought it was more consistent than that. So yeah, this isn't a canonical anything.

Reply:
Thanks for the survey!  So .. how should we document the current semantics of this? As "an opaque human readable string"?  Basically you can't rely on it being anything in particular, other than a value related to the node 'somehow'.

Maybe we just own up to the current state, and say: "NodeHostName is not a useful value. It can be _any_ of DNS name, os.Hostname or node.Name - and the expected semantics are undefined.  We recommend using NodeInternalDNS or node.Name instead."

Reply:
Maybe I'm not getting right the english meaning of canonical , but I was thinking that  "canonical name is what the k8s admin, cluster provider, or whoever sets the names on the nodes want to use for the primary name of the node, independently on how it gets that name ... in DNS the CNAME is a record and there is not rule too ...